pep-0xxx.rst

```diff
@@ -0,0 +1,1149 @@
+            """
+            Create a new context object from a given TLS configuration.
+            """
+
+        @property
+        @abstractmethod
+        def configuration(self) -> TLSConfiguration:
+            """
+            Returns the TLS configuration that was used to create the context.
+            """
+
+
+    class ClientContext(_BaseContext):
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket,
+                        auto_handshake=True: bool,
+                        server_hostname=None: Optional[str]) -> TLSWrappedSocket:
+            """
+            Wrap an existing Python socket object ``socket`` and return a
+            ``TLSWrappedSocket`` object. ``socket`` must be a ``SOCK_STREAM``
+            socket: all other socket types are unsupported.
+
+            The returned SSL socket is tied to the context, its settings and
+            certificates.
+
+            The parameter ``auto_handshake`` specifies whether to do the SSL
+            handshake automatically after doing a ``socket.connect()``, or
+            whether the application program will call it explicitly, by
+            invoking the ``TLSWrappedSocket.do_handshake()`` method. Calling
+            ``TLSWrappedSocket.do_handshake()`` explicitly gives the program
+            control over the blocking behavior of the socket I/O involved in
+            the handshake.
+
+            The optional parameter ``server_hostname`` specifies the hostname
+            of the service which we are connecting to. This allows a single
+            server to host multiple SSL-based services with distinct
+            certificates, quite similarly to HTTP virtual hosts.
+            """
+
+        @abstractmethod
+        def wrap_buffers(self, incoming: Any, outgoing: Any,
+                         server_hostname=None: Optional[str]) -> TLSWrappedBuffer:
+            """
+            Wrap a pair of buffer objects (``incoming`` and ``outgoing``) to
+            create an in-memory stream for TLS. The SSL routines will read data
+            from ``incoming`` and decrypt it, and write encrypted data to
+            ``outgoing``.
+
+            The ``server_hostname`` parameter has the same meaning as in
+            ``wrap_socket``.
+            """
+
+
+    class ServerContext(_BaseContext):
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket,
+                        auto_handshake=True: bool) -> TLSWrappedSocket:
+            """
+            Wrap an existing Python socket object ``socket`` and return a
+            ``TLSWrappedSocket`` object. ``socket`` must be a ``SOCK_STREAM``
+            socket: all other socket types are unsupported.
+
+            The returned SSL socket is tied to the context, its settings and
+            certificates.
+
+            The parameter ``auto_handshake`` specifies whether to do the SSL
+            handshake automatically after doing a ``socket.connect()``, or
```

Comment:
This doesn't seem like it makes sense here, you don't ever call ``socket.connect()`` on a server side socket do you? You call  ``listen()``, ``bind()``, ``accept()``, but no ``connect()`` afaik. This should probably be changed to ``after a ``socket.accept()``?